File: ios/Yomiyasu/Core/Settings/RandomCriteriaStore.swift
import Foundation

struct RandomCriteria: Codable, Equatable, Sendable {
    var genre: String?
    var author: String?
    var status: SerieStatus?
    var firstLetter: String?
    var minDifficulty: Int?
    var maxDifficulty: Int?
    var readprogress: ProgressFilter?
    var readlistOnly = false

    init(query: SeriesQuery) {
        genre = query.genre
        author = query.author
        status = query.status
        firstLetter = query.firstLetter
        minDifficulty = query.minDifficulty
        maxDifficulty = query.maxDifficulty
        readprogress = query.readprogress
        readlistOnly = query.readlistOnly
    }

    func applying(to query: SeriesQuery) -> SeriesQuery {
        var result = query
        result.genre = genre
        result.author = author
        result.status = status
        result.firstLetter = firstLetter
        result.minDifficulty = minDifficulty
        result.maxDifficulty = maxDifficulty
        result.readprogress = readprogress
        result.readlistOnly = readlistOnly
        return result
    }
}

@MainActor
final class RandomCriteriaStore {
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func save(_ criteria: RandomCriteria, variant: LibraryVariant) {
        guard let data = try? JSONEncoder().encode(criteria) else { return }
        defaults.set(data, forKey: Self.key(variant))
    }

    func criteria(variant: LibraryVariant) -> RandomCriteria? {
        guard let data = defaults.data(forKey: Self.key(variant)) else { return nil }
        return try? JSONDecoder().decode(RandomCriteria.self, from: data)
    }

    private static func key(_ variant: LibraryVariant) -> String {
        "randomCriteria.\(variant.rawValue)"
    }
}

File: android/app/src/main/java/es/manabe/yomiyasu/core/settings/RandomCriteriaStore.kt
package es.manabe.yomiyasu.core.settings

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import es.manabe.yomiyasu.core.models.LibraryVariant
import es.manabe.yomiyasu.core.models.ProgressFilter
import es.manabe.yomiyasu.core.models.SerieStatus
import es.manabe.yomiyasu.core.models.SeriesQuery
import kotlinx.coroutines.flow.first
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class RandomCriteria(
    val genre: String? = null,
    val author: String? = null,
    val status: String? = null,
    val firstLetter: String? = null,
    val minDifficulty: Int? = null,
    val maxDifficulty: Int? = null,
    val readprogress: String? = null,
    val readlistOnly: Boolean = false,
) {
    fun applyingTo(query: SeriesQuery): SeriesQuery = query.copy(
        genre = genre,
        author = author,
        status = status?.let { raw -> SerieStatus.entries.firstOrNull { it.rawValue == raw } },
        firstLetter = firstLetter,
        minDifficulty = minDifficulty,
        maxDifficulty = maxDifficulty,
        readprogress = readprogress?.let { raw ->
            ProgressFilter.entries.firstOrNull { it.rawValue == raw }
        },
        readlistOnly = readlistOnly,
    )

    companion object {
        fun from(query: SeriesQuery): RandomCriteria = RandomCriteria(
            genre = query.genre,
            author = query.author,
            status = query.status?.rawValue,
            firstLetter = query.firstLetter,
            minDifficulty = query.minDifficulty,
            maxDifficulty = query.maxDifficulty,
            readprogress = query.readprogress?.rawValue,
            readlistOnly = query.readlistOnly,
        )
    }
}

@Singleton
class RandomCriteriaStore @Inject constructor(
    private val dataStore: DataStore<Preferences>,
) {
    private val json = Json

    suspend fun save(criteria: RandomCriteria, variant: LibraryVariant) {
        dataStore.edit { prefs -> prefs[key(variant)] = json.encodeToString(criteria) }
    }

    suspend fun criteria(variant: LibraryVariant): RandomCriteria? {
        val raw = dataStore.data.first()[key(variant)] ?: return null
        return runCatching { json.decodeFromString<RandomCriteria>(raw) }.getOrNull()
    }

    private fun key(variant: LibraryVariant) = stringPreferencesKey("randomCriteria.${variant.rawValue}")
}
